src/action.py

```diff
@@ -1,30 +1,79 @@
 
 import glob
 import click
-import os
+import re
+
+from sys import exit
+from os import listdir
+from os.path import join, dirname, relpath, isfile, isdir
 from models.config import Config
 from collections import defaultdict
 from itertools import chain
 
+#TODO DRY up the file leader - only define '/var/lib/adminware/tools' once
 class ClickGlob:
-    def __glob_configs(namespace):
-        parts = ['/var/lib/adminware/tools',
-                 namespace, '*/config.yaml']
-        paths = glob.glob(os.path.join(*parts))
-        return list(map(lambda x: Config(x), paths))
+    def __glob_dirs(namespace):
+        parts = ['/var/lib/adminware/tools', namespace, '*']
+        paths = glob.glob(join(*parts))
+        return paths
+
+    # decorator for __glob_dir to get all valid paths (ending in config.yaml) at once
+    def __glob_all(globbing_func):
+        def __wrapper(*args, **kwargs):
+            def __sub_wrapper(*args, **kwargs):
+                paths = globbing_func(*args)
+                for path in paths:
+                    if isfile('{}/config.yaml'.format(path)):
+                        kwargs['all_paths'] += [join(path, 'config.yaml')]
+                    else:
+                        namespace = path[len('/var/lib/adminware/tools/'):]
+                        __sub_wrapper(namespace, **kwargs)
+                return kwargs['all_paths']
+
+            kwargs['all_paths']=[]
```

Comment:
This line is quite unintuitive  in it's purpose. It shadows the `all_paths` variable in name but isn't used again within this current scope.

Instead it looks to be used as a collector within `__sub_wrapper`. Is it possible to move this to somewhere up there?

Reply:
done here https://github.com/alces-software/adminware/pull/87/commits/be89624fbf4de21d3f15fe67aaae6a0530be8a56

Reply:
This commit still passes in the empty `all_paths` array. Why do we need this as an input?

Reply:
Well otherwise it'll throw a KeyError during the execution of `__sub_wrapper` & it can't be declared as an empty list _within_  `__sub_wrapper` because it'll be reset for each loop 

However if i just change `__glob_all` to it's own method instead of a decorator as per  https://github.com/alces-software/adminware/pull/87#discussion_r224516962 this stops being an issue